tests/unit/test_charm.py

```diff
@@ -458,71 +351,64 @@ def test_given_upf_installed_when_remove_then_snap_removed(self, patched_snap_ca
         self.harness.set_leader(True)
         self.harness.charm.on.install.emit()
 
         self.harness.evaluate_status()
 
-        self.assertEqual(
-            self.harness.model.unit.status,
-            ops.ActiveStatus(),
-        )
-
-
-class TestCharmInitialisation(unittest.TestCase):
-    @patch("charm.UPFNetwork")
-    @patch("charm.Machine")
-    def setUp(self, patch_machine, patch_network):
-        self.mock_machine = MagicMock()
-        self.mock_machine.pull.return_value = ""
-        self.mock_process = MagicMock()
-        self.mock_process.wait_output.return_value = ("", "")
-        self.mock_machine.exec.return_value = self.mock_process
-        patch_machine.return_value = self.mock_machine
-        self.mock_upf_network = MagicMock()
-        self.mock_upf_network.get_invalid_network_interfaces.return_value = []
-        self.mock_upf_network.core_interface.get_ip_address.return_value = (
-            "192.168.250.3"
-        )
-        patch_network.return_value = self.mock_upf_network
+        assert self.harness.model.unit.status == ops.ActiveStatus()
+
+
+class TestCharmInitialisation:
+    patcher_upf_network = patch("charm.UPFNetwork")
+    patcher_machine = patch("charm.Machine")
+    patcher_snap_cache = patch("charm.SnapCache")
+
+    @pytest.fixture()
+    def setup(self):
+        mock_upf_network = MagicMock()
+        mock_upf_network.get_invalid_network_interfaces.return_value = []
+        mock_upf_network.core_interface.get_ip_address.return_value = "192.168.250.3"
+        mock_machine = MagicMock()
+        mock_machine.pull.return_value = ""
+        mock_process = MagicMock()
+        mock_process.wait_output.return_value = ("", "")
+        mock_machine.exec.return_value = mock_process
+        self.mock_upf_network = TestCharmInitialisation.patcher_upf_network.start()
+        self.mock_upf_network.return_value = mock_upf_network
+        self.mock_machine = TestCharmInitialisation.patcher_machine.start()
+        self.mock_machine.return_value = mock_machine
+        self.mock_snap_cache = TestCharmInitialisation.patcher_snap_cache.start()
+
+    @staticmethod
+    def teardown() -> None:
+        patch.stopall()
+
+    @pytest.fixture(autouse=True)
+    def create_harness(self, setup, request):
         self.harness = ops.testing.Harness(SdcoreUpfCharm)
-        self.addCleanup(self.harness.cleanup)
+        self.harness.set_model_name(name="whatever")
+        self.harness.set_leader(is_leader=True)
+        yield self.harness
+        self.harness.cleanup()
+        request.addfinalizer(self.teardown)
 
-    @patch("charm.SnapCache")
-    def test_given_invalid_iface_ip_config_when_config_changed_then_status_is_blocked(self, _):
-        self.harness.set_leader(True)
+    def test_given_invalid_iface_ip_config_when_config_changed_then_status_is_blocked(self):
         self.harness.update_config({"access-ip": "not an ip address"})
         self.harness.begin()
```

Comment:
I am just wondering why the `begin` is here

Reply:
[This PR](https://github.com/canonical/sdcore-upf-k8s-operator/pull/147) explains why it is necessary.